x/twap/api_test.go

```diff
@@ -896,12 +896,13 @@ func (s *TestSuite) TestGeometricTwapToNow_BalancerPool_Randomized() {
 			twap, err := app.TwapKeeper.GetGeometricTwapToNow(ctx, 1, denom0, denom1, oldTime)
 			s.Require().NoError(err)
 
-			osmomath.ErrTolerance{
-				MultiplicativeTolerance: osmomath.SmallestDec(),
+			compareResult := osmomath.ErrTolerance{
+				MultiplicativeTolerance: osmomath.SmallestBigDec().Dec(),
```

Comment:
we cannot set `MultiplicativeTolerance` like this: it effectively means that the multiplicative error tolerance is zero, since converting smallest big dec to dec would yield a 0 in Dec representation (since Dec truncates values after `DecPrecision`'s decimal). 

the current comparison logic treats `MultiplicativeTolerance` that is set to either nil or zero as an indication that we should not assert multiplicative tolerance: https://github.com/osmosis-labs/osmosis/blob/main/osmomath/binary_search.go#L71.

the reason it is passing, is because it does not check any difference: either additive (it is nil) nor multiplicative (it is zero)

Reply:
yeah, I checked ZeroDec and expel it from the test but forgot this will also have similar effect too. Thank you for the head up

Reply:
as I checked, the biggest spike would be "0.000000032318177561975767141009575914" and smallest spike that fail will be "0.000000000437394800044341685165135218". both will be bigger than `MultiplicativeTolerance ` and result in fail. Do you think multiply `SmallestDec ` so that this test pass is acceptable ?

Reply:
```suggestion
				MultiplicativeTolerance: osmomath.SmallestBigDec().Dec(),
```
to `osmomath.SmallestDec().Mul(osmomath.NewDec(32318177563))` and it will pass.

Reply:
@pysel I expect we can't hardcode any value. I'll come up with a calculation for this multiply value tmr.

Reply:
@pysel so round up `MultiplicativeTolerance ` and the test would pass. Do you want me spend more time to find the upper bound of `spotPrice` as well as `osmomath.BigDecFromDec(twap)` so that we have a definitive way to calculate this or `osmomath.NewDec(40000000000)` /  0.000000040000000000 is a acceptable  value for MultiplicativeTolerance ?

Reply:
I am honestly not sure if 4 * 10^-8 is an acceptable error in this case. Do you have some info about test cases that fail with smallest dec for multiplicative tolerance? since twap is computed over a random time period:
```
        // from test 
	elapsedTimeMs := sdkrand.RandIntBetween(r, 1, int(maxUint64>>1))
```
I suspect that, maybe, test cases that fail tend to have relatively small period over which twap is computed. hence, there might be not much price points over which twap is being calculated, so the approximation of an actual spot price is badly estimated. 

I think it is better to ask someone else on that - going to tag code owners of twap for this one:

cc: @nicolaslara @p0mvn @AlpinYukseloglu

Reply:
![image](https://github.com/osmosis-labs/osmosis/assets/63839106/c47cd919-beb1-4acb-b43d-44c0483053af)
I'm not sure if elapsedTimes is too small, here is the log

Reply:
@pysel if you want me to spend more time on figure what the upper bound of estimate price and actual price value, the reason/calculation behind if then just give me a head up.
I think for this test I should break down the range of spotPrice and estimateTwapPrice, and then we can find the final cause of this.

Reply:
I think it might not be worth the time, I think it would be more reasonable to wait for other devs to comment on this. still, I think it might be worth it to increase the lower bound of elapsed time in test by few orders of magnitude, maybe that would help, maybe that is a bad advice, not sure honestly